File: env_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Client to play multi-armed bandits problem against.
Many distribution of arms are supported, default to Bernoulli.

Usage:
    env_client.py [markovian | dynamic] [--port=<PORT>] [--host=<HOST>] [--speed=<SPEED>] <json_configuration>
    env_client.py (-h|--help)
    env_client.py --version

Options:
    -h --help   Show this screen.
    --version   Show version.
    markovian   Whether to use a Markovian MAB problem (default is simple MAB problems).
    dynamic     Whether to use a Dynamic MAB problem (default is simple MAB problems).
    --port=<PORT>   Port to use for the TCP connection [default: 10000].
    --host=<HOST>   Address to use for the TCP connection [default: 0.0.0.0].
    --speed=<SPEED>   Speed of emission in milliseconds [default: 1000].
"""
from __future__ import division, print_function  # Python 2 compatibility

__author__ = "Lilian Besson"
__version__ = "0.7"
version = "MAB environment client v{}".format(__version__)

import json
import socket
import time
from docopt import docopt

from Environment import MAB, MarkovianMAB, DynamicMAB
from Arms import *


#: Example of configuration to pass from the command line.
#: ``'{"arm_type": "Bernoulli", "params": (0.1, 0.5, 0.9)}'``
default_configuration = {
        "arm_type": "Bernoulli",
        "params": {
            (0.1, 0.5, 0.9)
        }
    }


def read_configuration_env(a_string):
    """ Return a valid configuration dictionary to initialize a MAB environment, from the input string."""
    obj = json.loads(a_string)
    assert isinstance(obj, dict) and "arm_type" in obj and "params" in obj, "Error: invalid string to be converted to a configuration object for a MAB environment."
    return obj


def send_message(sock, message):
    # Send data
    print("sending {!r}".format(message))
    sock.sendall(message)


def client(env, host, port, speed):
    """
    Launch an client that:

    - uses sockets to listen to input and reply
    - create a MAB environment from a JSON configuration (exactly like ``main.py`` when it reads ``configuration.py``)
    - then receives choice ``arm`` from the network, pass it to the MAB environment, listens to his ``reward = draw(arm)`` feedback, and sends this back to the network.
    """
    # Create a TCP/IP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    # Bind the socket to the port
    server_address = (host, port)
    print("starting up on {} port {}".format(*server_address))
    sock.connect(server_address)
    t = -1

    try:
        message = "Hi!".encode()
        print("\nSending first message = {!r}".format(message))
        send_message(sock, message)
        print("Sleeping for {} second(s)...".format(speed))
        time.sleep(speed)
        while True:
            t += 1
            arm = t % env.nbArms

            data = sock.recv(16)
            message = data.decode()
            print("\nData received: {!r}".format(message))
            arm = int(message)

            reward = env.draw(arm)
            message = str(reward)[:7].encode()
            print("Environment = {}, at time t = {}:".format(env, t))
            print("Sending random reward = {!r} from arm {} ...".format(message, arm))
            send_message(sock, message)
            print("Sleeping for {} second(s)...".format(speed))
            time.sleep(speed)
    finally:
        # Clean up the socket
        print("Closing socket...")
        sock.close()


def transform_str(params):
    """Like a safe exec() on a dictionary that can contain special values:

    - strings are interpreted as variables names (e.g., policy names) from the current ``globals()`` scope,
    - list are transformed to tuples to be constant and hashable,
    - dictionary are recursively transformed.
    """
    for (key, value) in params.items():
        if isinstance(value, dict):
            transform_str(value)
        if isinstance(value, list):  # unhashable
            value = tuple(value)
        try:
            if value in globals():
                params[key] = globals()[value]
        except TypeError:
            pass


def main(arguments):
    """
    Take arguments, construct the learning policy and starts the server.
    """
    host = arguments['--host']
    port = int(arguments['--port'])
    speed = float(arguments['--speed']) / 1000.0
    is_markovian = arguments['markovian']
    is_dynamic = arguments['dynamic']
    json_configuration = arguments['<json_configuration>']
    configuration = read_configuration_env(json_configuration)
    transform_str(configuration)
    if is_dynamic:
        env = DynamicMAB(configuration)
    elif is_markovian:
        env = MarkovianMAB(configuration)
    else:
        env = MAB(configuration)
    print("Using the environment: ", env)  # DEBUG
    print("Emitting regularly every", speed, "seconds.")  # DEBUG
    return client(env, host, port, speed)


if __name__ == '__main__':
    arguments = docopt(__doc__, version=version)
    # print("arguments =", arguments)  # DEBUG
    main(arguments)

File: SMPyBandits/env_client.cpp
// #! g++ -Wall -Iinclude -o env_client.exe include/docopt.cpp env_client.cpp
/**
    C++ client, using sockets, to simulate a MAB environment.
    So far, only Bernoulli arms are supported.

    - Author: Lilian Besson
    - License: MIT License (https://lbesson.mit-license.org/)
    - Date: 28-07-2017
    - Online: https://smpybandits.github.io/
    - Reference: http://www.binarytides.com/code-a-simple-socket-client-class-in-c/
*/

// Include libraries
#include <arpa/inet.h>  // inet_addr
#include <chrono>       // milliseconds
#include <cstdlib>      // rand
#include <docopt.h>     // docopt command line parser
#include <iostream>     // streams, <<, >>
#include <netdb.h>      // hostent
#include <stdio.h>      // printf
#include <string.h>     // strlen
#include <string>       // string
#include <sys/socket.h> // socket
#include <thread>       // sleep

// No need for std::printf, std::string etc
using namespace std;

/**
    TCP Client class
*/
class tcp_client {
    private:
        int sock;
        string address;
        int port;
        struct sockaddr_in server;

    public:
        tcp_client();
        bool conn(string, int);
        bool send_data(string data);
        string receive(int);
};

/**
    Default initializer
*/
tcp_client::tcp_client() {
    sock = -1;
    port = 0;
    address = "";
}

/**
    Connect to a host on a certain port number
*/
bool tcp_client::conn(string address, int port) {
    const char *c_address;
    c_address = address.c_str();

    // create socket if it is not already created
    if (sock == -1) {
        // Create socket
        sock = socket(AF_INET, SOCK_STREAM, 0);
        if (sock == -1) {
            perror("Could not create socket");
        }
        printf("Socket created\n");
    } else {
        /* OK , nothing */
    }

    // setup address structure
    if (inet_addr(c_address) < 0) {
        struct hostent *he;
        struct in_addr **addr_list;

        // resolve the hostname, its not an ip address
        if ((he = gethostbyname(c_address)) == NULL) {
            // gethostbyname failed
            herror("gethostbyname");
            printf("Failed to resolve hostname for '%s'... Try again please.\n",
                    c_address);
            return false;
        }

        // Cast the h_addr_list to in_addr , since h_addr_list also has the ip
        // address in long format only
        addr_list = (struct in_addr **)he->h_addr_list;

        for (int i = 0; addr_list[i] != NULL; i++) {
            server.sin_addr = *addr_list[i];
            printf("Address '%s' resolved to '%s'...\n", c_address, inet_ntoa(*addr_list[i]));
            break;
        }
    }
    // plain ip address
    else {
        server.sin_addr.s_addr = inet_addr(c_address);
    }

    server.sin_family = AF_INET;
    server.sin_port = htons(port);

    // Connect to remote server
    if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
        perror("Connect failed. Error!");
        return false;
    }

    printf("Connected to '%s' with port '%d' !\n", c_address, port);
    return true;
}

/**
    Send data to the connected host
*/
bool tcp_client::send_data(string data) {
    // Send some data
    if (send(sock, data.c_str(), strlen(data.c_str()), 0) < 0) {
        perror("Send failed : ");
        return false;
    }
    printf("\nData '%s' successfully sent!\n", data.c_str());
    return true;
}

/**
    Receive data from the connected host
*/
string tcp_client::receive(int size = 4) {
    char buffer[size];
    string reply;

    // Receive a reply from the server
    if (recv(sock, buffer, sizeof(buffer), 0) < 0) {
        printf("recv failed...");
    }

    reply = buffer;
    return reply;
}

// Macro to have a random float in [0, 1)
#define random_float() (rand() / static_cast<float>(RAND_MAX))

/**
    Draw one sample from a Bernoulli distribution of a certain mean.
*/
float bernoulli_draw(float mean) {
    // send some data, random in [0, 1]
    if (random_float() < mean) {
        return 1;
    } else {
        return 0;
    }
}

/**
    Infinite loop, sending random rewards on the asked arm.

    - create the socket and connect,
    - continuously read the socket for a channel id number #i,
    - generate a random reward for mean = mu[i],
    - send back that reward to the socket.
*/
int loop(
    string address,
    int port,
    vector<float> means,
    int milli_sleep = 2000
) {
    srand(time(0)); // use current time as seed for random generator

    tcp_client c;
    string received;
    float reward;
    int channel;

    c.conn(address, port); // connect to host

    // send some data, just a stupid useless handcheck
    c.send_data("Hi from env_client.exe !");

    while (true) { // receive and echo reply
        try {
            received = c.receive();
            try {
                channel = stoi(received);
            } catch (const invalid_argument &) {
                channel = 0;
            }
            printf("\nReceived '%s' = channel #'%d'...", received.c_str(), channel);
            reward = bernoulli_draw(means[channel]);
            c.send_data(to_string(reward));
        } catch (const invalid_argument &) {
            printf("\nReceived something not correctly understood by stoi(), no problem we continue.");
        } catch (const runtime_error &) {
            printf("\nRuntime error in the loop, no issue we continue.");
        }
        this_thread::sleep_for(chrono::milliseconds(milli_sleep));
    };

    return 0; // done
}

/**
    Convert a string, read from the cli, to a vector of float number.
    In Python, that would be map(float, arr.split(',')), but it takes 15 lines
    here. Yay! Finally debugged, OK.
*/
vector<float> array_from_str(string arr) {
    // cout << "arr = " << arr << endl;  // DEBUG
    uint nb = 1;
    uint size_arr = arr.size();
    // cout << "size_arr = " << size_arr << endl;  // DEBUG
    // first, find nb
    uint index = 0;
    uint found;
    while (true) {
        found = arr.find(',', index);
        // cout << "\nfound = " << found << endl;  // DEBUG
        // cout << "index = " << index << endl;  // DEBUG
        if ((found == string::npos) || (found >= size_arr)) {
            break;
        } else {
            nb += 1;
            index = found + 1;
        }
    }
    // cout << "nb = " << nb << endl;  // DEBUG

    // allocate vector
    vector<float> means;
    // cout << "\nSecond step...\n";  // DEBUG

    // then iterate on the string, convert and store
    uint i = 0;
    index = 0;
    while (true) {
        found = arr.find(',', index);
        // cout << "\nfound = " << found << endl;  // DEBUG
        // cout << "index = " << index << endl;  // DEBUG
        if (i > nb) {
            break;
        } else {
            // cout << "reading = " << stof(arr.substr(index, found)) << endl;  // DEBUG
            means.push_back(stof(arr.substr(index, found)));
            // cout << "means[i] = " << means[i] << endl;  // DEBUG
            index = found + 1;
            i += 1;
            // cout << "i = " << i << endl;  // DEBUG
        }
    }
    return means;
}

/**
    documentation for the cli generated with docopt
*/
static const char USAGE[] =
    R"(C++ Client to play multi-armed bandits problem against.

Usage:
    env_client.exe [--port=<PORT>] [--host=<HOST>] [--speed=<SPEED>] [<bernoulli_means>]
    env_client.exe (-h|--help)
    env_client.exe --version

Options:
    -h --help   Show this screen.
    --version   Show version.
    --port=<PORT>   Port to use for the TCP connection [default: 10000].
    --host=<HOST>   Address to use for the TCP connection [default: 0.0.0.0].
    --speed=<SPEED>   Speed of emission in milliseconds [default: 1000].
)";

/**
    Main function, parsing the cli arguments with docopt::docopt and calling
    loop() with the good arguments.
*/
int main(
    int argc,
    const char **argv
) {
    string address;
    long port, speed;
    vector<float> default_means = {0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9};
    vector<float> means;

    // parse the cli arguments, magically with docopt::docopt
    map<string, docopt::value> args =
        docopt::docopt(USAGE, {argv + 1, argv + argc},
                        true,                             // show help if requested
                        "MAB environment C++ client v0.1" // version string
        );

    address = args["--host"].asString();
    port = args["--port"].asLong();
    speed = args["--speed"].asLong();
    if (args["<bernoulli_means>"].isString()) {
        printf("Bernoulli means = '%s'\n", args["<bernoulli_means>"].asString());
        means = array_from_str(args["<bernoulli_means>"].asString());
    } else {
        means = default_means;
    }

    cout << "- address = " << address << endl; // DEBUG
    cout << "- port = " << port << endl;       // DEBUG
    cout << "- speed = " << speed << endl;     // DEBUG
    for (uint i = 0; i < means.size(); i++) {
        cout << "- means[" << i << "] = " << means[i] << endl; // DEBUG
    }
    cout << endl << "Calling loop... starting..." << endl; // DEBUG
    return loop(address, port, means, speed);
}